### PR TITLE
Scale 12 OCS workers, 36 OSDs and 6000 PVCs attached to PODs and perform tests

### PR DESCRIPTION
**Pull request by ramkiperiy**
To create 12 OCS workers, 36 OSDs, 6000 PVCs
Further testing re-spin, reboot and shutdown of pods, nodes

Signed-off-by: Ramakrishnan <rperiyas@redhat.com>

### CODE REVIEW

**Comment by ramkiperiy** [comment]
> PR validation
> 
> Cluster Name:
> Cluster Configuration:
> PR Test Suite: scale
> PR Test Path: tests/e2e/scale/test_scale_12_OCS_worker_nodes_and_6000_PVCs.py
> Additional Test Params:
> OCP VERSION: 4.8
> OCS VERSION: 4.8
> 
> [Job UNSTABLE](https://ocs4-jenkins-csb-ocsqe.apps.ocp4.prod.psi.redhat.com/job/qe-trigger-test-pr/1064//testReport) (some or all tests failed).

Failure Analysis:
1. tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_scale_node_and_capacity  
    Test failed in teardown, due to cluster health no OK, i.e. Scale of nodes and OSD"s are successful, will add validation for cluster to be in good state.
2. tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_scale_pvcs_pods
    Test failed for PG_balancer function, scaling of 6000 PVCs and 300 pods are successful, will see how the pg_validation can be done/changed.
3. tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[mgr]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[mon]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[osd]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[mds]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[cephfsplugin]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[rbdplugin]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[rbdplugin]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[cephfsplugin_provisioner]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[rbdplugin_provisioner]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_respin_ceph_pods[operator]
   Test Failed due to previous test, i.e. in previous test after creation of 6000 pvc's data will be collected in a file for validating in this tc, since above test failed the file is not created so all above tests failed. but it did re-spin of the respective pods which is successful
4. tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_reboot_nodes[worker]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_reboot_nodes[master]
tests.e2e.scale.test_scale_12_OCS_worker_nodes_and_6000_PVCs.TestAddNode.test_shutdown_nodes
  Will be removing above 3 tests since for now there is no code.

Overall changes to be done are listed below.
1. Add ceph health validation in each testcase
2. Add validation for scale_data file available
3. Add code to skip tests if previous tests failed (in-case like dependent tests)
4. Remove/add code for Reboot and shutdown tests.

File: tests/e2e/scale/test_scale_12_OCS_worker_nodes_and_6000_PVCs.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+# Scale data file
+log_path = ocsci_log_path()
+SCALE_DATA_FILE = f"{log_path}/scale_data_file.yaml"
+
+
+@ignore_leftovers
+@scale
+@skipif_aws_i3
+@skipif_bm
+@skipif_external_mode
```
Comment (vasukulkarni) [inline tests/e2e/scale/test_scale_12_OCS_worker_nodes_and_6000_PVCs.py:32]: can we skip for ibm?

File: tests/e2e/scale/test_scale_12_OCS_worker_nodes_and_6000_PVCs.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+            if len(existing_ocs_worker_list) < expected_worker_count:
+                scale_worker_count = expected_worker_count - len(
+                    existing_ocs_worker_list
+                )
+                scale_lib.scale_ocs_node(node_count=scale_worker_count)
+
+            # Check existing OSD count and add OSDs if required
+            if existing_deviceset_count < expected_deviceset_count:
+                add_deviceset_count = (
+                    expected_deviceset_count - existing_deviceset_count
```
Comment (PrasadDesala) [inline tests/e2e/scale/test_scale_12_OCS_worker_nodes_and_6000_PVCs.py:69]: In LSO based cluster the the replica count is 1, does these lines of code take that also into factor while calculating the add_deviceset_count

Comment (ramkiperiy) [reply tests/e2e/scale/test_scale_12_OCS_worker_nodes_and_6000_PVCs.py:69]: @PrasadDesala That's good point, let me add skipif_lso marker, anyway we don't have support to run some scale tests(scale capacity, node tests) in UPI cluster.

**Comment by ramkiperiy** [comment]
> PR validation
> 
> Cluster Name:
> Cluster Configuration:
> PR Test Suite: scale
> PR Test Path: tests/e2e/scale/test_scale_12_OCS_worker_nodes_and_6000_PVCs.py
> Additional Test Params:
> OCP VERSION: 4.8
> OCS VERSION: 4.8
> 
> [Job UNSTABLE](https://ocs4-jenkins-csb-ocsqe.apps.ocp4.prod.psi.redhat.com/job/qe-trigger-test-pr/1110//testReport) (some or all tests failed).

All tests passed but there is a failure in teardown due to resource leftover which is expected, i.e. should add @ignore_leftover marker, there are restart of ceph pods in the tests, which is raising this issue.
image

**Comment by ramkiperiy** [comment]
> PR validation
> 
> Cluster Name:
> Cluster Configuration:
> PR Test Suite: scale
> PR Test Path: tests/e2e/scale/test_scale_12_OCS_worker_nodes_and_6000_PVCs.py
> Additional Test Params:
> OCP VERSION: 4.8
> OCS VERSION: 4.8
> 
> [Job PASSED](https://ocs4-jenkins-csb-ocsqe.apps.ocp4.prod.psi.redhat.com/job/qe-trigger-test-pr/1112//testReport).

image

**Review by vasukulkarni — APPROVED** [review]

**Comment by ramkiperiy** [comment]
PR needs re-base once #4353 & #4565 are merged.

**Review by vasukulkarni — APPROVED** [review]

**Review by wusui — COMMENTED** [review]
So @ignore_leftovers is needed for this?  How does this use differ from using ignore_leftovers for https://github.com/red-hat-storage/ocs-ci/pull/4133

**Comment by ramkiperiy** [comment]
> So @ignore_leftovers is needed for this? How does this use differ from using ignore_leftovers for #4133

am performing Ceph and operator pod restarts, due to restarts those pods name will change which pytest identifies as leftovers, most of the pod restart tests will have ignore_leftover marker to avoid the failure in teardown.

**Review by Avilir — APPROVED** [review]
LGTM

**Review by keesturam — APPROVED** [review]

**Pull request merged**